### ISSUE TITLE
Error creating Lambda event source mapping: Function does not exist

### ISSUE THREAD

**Issue by kalpaitch** [aws/chalice#1558]
Another terraform packaging issue here, when using additional resources (sqs queues or s3 buckets) that are created outside of the chalice application, but packaged and deployed together with terraform. There seems to be a missing dependency between the lambda permission and the lambda function. Additionally there seems no way to define the dependency between the bucket notification and the bucket itself:
```
"aws_lambda_function": {
    "lambda_function_name": {
        "function_name": "lambda_function_name",
        "role": "${aws_iam_role.default-role.arn}",
        ...
    },
"aws_s3_bucket_notification": {
    "project-bucket-name_notify": {
        "bucket": "project-bucket-name",
        "lambda_function": [
            {
                "events": [
                    "s3:ObjectCreated:*"
                ],
                "lambda_function_arn": "${aws_lambda_function.lambda_function_name.arn}"
            }
        ]
    }
},
"aws_lambda_permission": {
    "lambda_function_name-s3event": {
        "action": "lambda:InvokeFunction",
        "function_name": "lambda_function_name",
        "source_arn": "arn:*:s3:::project-bucket-name"
        ...
    }...
}
```
&&
```
resource "aws_s3_bucket" "project_bucket" {
    bucket = "project-bucket-name"
    ...
}
```

This produces the following errors on first deploy (not on subsequent deploys):
```
Error: Error putting S3 notification configuration: InvalidArgument: Unable to validate the following destination configurations
	status code: 400...

Error: Error creating Lambda event source mapping: InvalidParameterValueException: Function does not exist
{
  RespMetadata: {
    StatusCode: 400,...
  },
  Message_: "Function does not exist",...
}
```

Adding a dependency to the permission resolves the error `Error creating Lambda event source mapping`:
```
"aws_lambda_permission": {
    "read_mail-s3event": {
        "action": "lambda:InvokeFunction",
        "function_name": "lambda_function_name",
        "source_arn": "arn:*:s3:::project-bucket-name"
        "depends_on": ["aws_lambda_function.lambda_function_name"]
    },
}
```

**Comment by kalpaitch** [comment]
Related to my other issue https://github.com/aws/chalice/issues/1557 in relation to the dependency errors.

**Comment by kalpaitch** [comment]
Equally being able to depend on the bucket is quite important, as per the example in https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket_notification#add-notification-configuration-to-lambda-function:

```
"aws_s3_bucket_notification": {
    "project-bucket-name_notify": {
        "bucket": "project-bucket-name",
        "lambda_function": [
            {
                "events": [
                    "s3:ObjectCreated:*"
                ],
                "lambda_function_arn": "${aws_lambda_function.lambda_function_name.arn}"
            }
        ],
        "depends_on": ["aws_lambda_permission.project_bucket"]
    }
}
```

**Comment by jamesls** [comment]
Not sure if this is related, but your existing config shouldn't work due to (https://github.com/aws/chalice/issues/1508), which was fixed in the latest release (1.21.2).  Terraform should be complaining about the `source_arn` being invalid so I'm curious how it was able to make it past that validation.  I confirmed that I'm able to deploy an app with an S3 bucket created as part of the terraform deployment with the latest chalice/terraform.

I wonder if an older version of terraform isn't waiting for the resource to finish creating before updating its permissions.   Can you check if you're running into these issues on the latest version of terraform and chalice?
